### PR TITLE
Resolve ReadTimeout errors in C3 client with retry mechanism

### PR DESCRIPTION
**Pull request by nadzyah**
The C3 data imports keeps failing during Docker builds due to ReadTimeout exceptions when the `certification.canonical.com` server is slow to respond.

## Implementation

The solution is:
- Adding exponential backoff retry logic (5 retries, 2-60 second delays)
- Implementing dual retry strategy: urllib3.Retry for HTTP errors + custom logic for timeouts
- Enhancing error handling for connection issues and server errors
- Adding comprehensive logging for retry attempts

## Testing

I've added 11 new test cases covering retry scenarios, edge cases, and pagination

### CODE REVIEW

File: server/hwapi/external/c3/client.py [hunk 1]
```diff
@@ -44,6 +47,77 @@ class C3Client:
...
+
+    def _make_request_with_retries(
+        self, url: str, max_retries=5, base_delay=2, max_delay=60
+    ):
+        """Make HTTP request with custom retry logic for timeout errors"""
+        for attempt in range(max_retries):
+            try:
+                logger.debug(
+                    f"Attempting request to {url} (attempt {attempt + 1}/{max_retries})"
+                )
```
Comment (pedro-avalos) [inline server/hwapi/external/c3/client.py:79]: It's recommended to not form the string in the log calls for (somewhat minor) efficiency gains:

```suggestion
                logger.debug(
                    "Attempting request to %s (attempt %d/%d)", url, attempt + 1, max_retries
                )
```

File: server/hwapi/external/c3/client.py [hunk 2]
```diff
@@ -44,6 +47,77 @@ class C3Client:
...
+                return response
+
+            except (
+                requests.exceptions.ReadTimeout,
+                requests.exceptions.ConnectTimeout,
+                requests.exceptions.ConnectionError,
+            ) as e:
+
+                if attempt == max_retries - 1:  # Last attempt
+                    logger.error(f"Failed to fetch {url} after {max_retries} attempts")
```
Comment (pedro-avalos) [inline server/hwapi/external/c3/client.py:91]:
```suggestion
                    logger.error("Failed to fetch %s after %d attempts", url, max_retries)
```

File: server/hwapi/external/c3/client.py [hunk 3]
```diff
@@ -44,6 +47,77 @@ class C3Client:
...
+                    raise e
+
+                # Calculate delay with exponential backoff
+                delay = min(base_delay * (2**attempt), max_delay)
+                logger.warning(
+                    f"Request to {url} failed (attempt {attempt + 1}/{max_retries}): {str(e)}. "
+                    f"Retrying in {delay} seconds..."
+                )
```
Comment (pedro-avalos) [inline server/hwapi/external/c3/client.py:99]:
```suggestion
                logger.warning(
                    "Request to %s failed (attempt %d/%d): %s. Retrying in %d seconds...",
                    url,
                    attempt + 1,
                    max_retries,
                    e,
                    delay,
                )
```

File: server/hwapi/external/c3/client.py [hunk 4]
```diff
@@ -44,6 +47,77 @@ class C3Client:
...
+                time.sleep(delay)
+
+            except requests.exceptions.HTTPError as e:
+                # For HTTP errors, check if it's worth retrying
+                if e.response.status_code in [429, 500, 502, 503, 504]:
```
Comment (pedro-avalos) [inline server/hwapi/external/c3/client.py:104]: I think this is a little more readable, let me know what you think:


```suggestion
                status_code = HTTPStatus(e.response.status_code)
                if status_code.is_server_error or status_code == HTTPStatus.TOO_MANY_REQUESTS:
```

The above needs the following import:

```python
from http import HTTPStatus
```

File: server/hwapi/external/c3/client.py [hunk 5]
```diff
@@ -44,6 +47,77 @@ class C3Client:
...
+                    if attempt == max_retries - 1:
+                        logger.error(
+                            f"Failed to fetch {url} after {max_retries} attempts"
+                        )
```
Comment (pedro-avalos) [inline server/hwapi/external/c3/client.py:108]:
```suggestion
                        logger.error(
                            "Failed to fetch %s after %d attempts", url, max_retries
                        )
```

File: server/hwapi/external/c3/client.py [hunk 6]
```diff
@@ -44,6 +47,77 @@ class C3Client:
...
+                        raise e
+
+                    delay = min(base_delay * (2**attempt), max_delay)
+                    logger.warning(
+                        f"HTTP error {e.response.status_code} for {url} "
+                        f"(attempt {attempt + 1}/{max_retries}). Retrying in {delay} seconds..."
+                    )
```
Comment (pedro-avalos) [inline server/hwapi/external/c3/client.py:115]:
```suggestion
                    logger.warning(
                        "HTTP error %d for %s (attempt %d/%d). Retrying in %d seconds...",
                        status_code,
                        url,
                        attempt + 1,
                        max_retries,
                        delay,
                    )
```

File: server/hwapi/external/c3/client.py [hunk 7]
```diff
@@ -44,6 +47,77 @@ class C3Client:
...
+                    time.sleep(delay)
+                else:
+                    # Don't retry for client errors (4xx except 429)
+                    logger.error(f"Non-retryable HTTP error for {url}: {e}")
```
Comment (pedro-avalos) [inline server/hwapi/external/c3/client.py:119]:
```suggestion
                    logger.error("Non-retryable HTTP error for %s: %s", url, e)
```

File: server/hwapi/external/c3/client.py [hunk 8]
```diff
@@ -95,25 +169,32 @@ def _import_from_c3(self, url: str, loader: Callable, resp_model: Type[BaseModel
...
+            # Use retry logic for each request
+            response = self._make_request_with_retries(next_url)
             response_json = response.json()
+
             if counter == 0:
-                # since count is always the same, update total only the first time
+                # Since count is always the same, update total only the first time
                 # we retrieve the data from C3
                 total = response_json["count"]
+                logger.info(f"Total items to process: {total}")
```
Comment (pedro-avalos) [inline server/hwapi/external/c3/client.py:185]:
```suggestion
                logger.info("Total items to process: %d", total)
```

**Review by pedro-avalos — CHANGES_REQUESTED** [review]
Love it; I've left some suggestions regarding logging and the use of magic numbers for HTTP Status codes.

**Review by pedro-avalos — APPROVED** [review]
LGTM, thanks for addressing the comments!
